### PR TITLE
Multi Layout View

### PR DESCRIPTION
**Pull request by vixalien**
Utilize libadwaita's shiny new MultiLayoutView

Should merge when `Adw.MultiLayoutView` is available in GNOME SDK master.

### CODE REVIEW

File: src/components/player/now-playing/view.ts [hunk 1]
```diff
@@ -1,365 +1,138 @@
...
-      );
-
-      return true;
-    }
+  set show_mini(show: boolean) {
+    this._multi_layout_view.layout_name = show ? "mini" : "full";
+    this._headerbar.show_end_title_buttons =
+      this._headerbar
+        .show_start_title_buttons =
+        show ? true : false;
```
Comment (DaPigGuy) [inline src/components/player/now-playing/view.ts:66]:
```suggestion
        show;
```

**Comment by DaPigGuy** [comment]
You've probably already seen this, but `AdwBottomSheet` has been made [public](https://gitlab.gnome.org/GNOME/libadwaita/-/merge_requests/1154). Might be nice to have with the "mini" items.

**Comment by vixalien** [comment]
Ah nice. That widget is interesting, and I will definitely use it.

What do you mean by "mini" widgets?

**Comment by DaPigGuy** [comment]
> What do you mean by "mini" widgets?

Sorry, I meant the detail items (queue, lyrics, related) when the window is mini

**Comment by vixalien** [comment]
Ah, I see. I will be updating in about a week (when the changes are available in the GNOME master)

**Comment by DaPigGuy** [comment]
It seems to already be available on org.gnome.Platform/master from gnome-nightly, as I can use bottom sheets in the Adwaita Demo app

**Comment by vixalien** [comment]
Oh cool, let me test that. I will work on updating this merge request, but it will be mergeable in a few days after [the gnome master image we use in CI](https://github.com/andyholmes/flatter/pkgs/container/flatter%2Fgnome) is updated.

**Comment by vixalien** [comment]
> It seems to already be available on org.gnome.Platform/master from gnome-nightly, as I can use bottom sheets in the Adwaita Demo app

On the other hand are you positive about this? I cannot see the bottom sheets in the Adwaita Demo app

**Comment by vixalien** [comment]
how did you launch it?

**Comment by DaPigGuy** [comment]
> how did you launch it?

Just from the app grid

**Comment by vixalien** [comment]
> Just from the app grid

hmm it should usually not be possible to do that as Adwaita Demo's  desktop file has `NoDisplay=true`. But anyway I think it's because of CDN propagation.

**Comment by DaPigGuy** [comment]
> > Just from the app grid
> 
> hmm it should usually not be possible to do that as Adwaita Demo's desktop file has `NoDisplay=true`. But anyway I think it's because of CDN propagation.

The Flatpak removes the `NoDisplay` [here](https://gitlab.gnome.org/GNOME/libadwaita/-/blob/main/demo/org.gnome.Adwaita1.Demo.json?ref_type=heads#L50)

**Comment by vixalien** [comment]
Ah, you installed adwaita demo itself. As you can see in the manifest, it builds `libadwaita` itself, so it is always updated. The `org.gnome.Platform//master` runtime gets updated a bit later, usually less than a week.

**Comment by vixalien** [comment]
On this case though, should the bottomsheet be the bottom player, and expand into the player controls?

**Comment by vixalien** [comment]
I have a better idea. I think the whole "now playing view" should be the bottom sheet. Then on mobile, as the user switches from the now playing view to the details (lyrics/queue/related), a new page would be switched to in the bottom sheet.

However, I think bottom sheets should never be used on desktop.

**Comment by DaPigGuy** [comment]
> However, I think bottom sheets should never be used on desktop.

Absolutely agreed.

> Then on mobile, as the user switches from the now playing view to the details (lyrics/queue/related), a new page would be switched to in the bottom sheet.

So something like below?

* AdwMultiLayoutView
  * AdwLayout "mobile"
    * AdwBottomSheet
      * sheet:
        * AdwViewStack +  (lyrics/queue/related tab buttons, click -> [set view to "details_stack", set view of "details_stack"])
          * now_playing
          * "details_stack"
      * bar: MiniPlayer
  * AdwLayout "desktop"
    * AdwOverlaySplitView
       * sidebar:  (lyrics/queue/related tab buttons, click -> set view of "details_stack") + "details"
       * content: "now_playing"
  * AdwLayoutSlot "now_playing"
    * Now Playing
  * AdwLayoutSlot "details"
    * AdwViewStack "details_stack"
      * Lyrics
      * Queue
      * Related

**Comment by vixalien** [comment]
something along those lines... I will try to create something and we can do feedback on the changes

**Comment by vixalien** [comment]
The new changes are here. Based [on this discussion](https://gitlab.gnome.org/GNOME/libadwaita/-/issues/728#note_1836501), this is how it works

#### On Desktop

There is a sidebar showing queue, lyrics, related

image

#### On Mobile

There is a bottom sheet and you can view the queue, lyrics, related, etc.. inline

image

image

Your feedback is appreciated

File: src/components/player/now-playing/details/lyrics.ts [hunk 2]
```diff
@@ -89,34 +89,33 @@ export class LyricsView extends Gtk.Stack {
...
+
+    this.set_visible_child(this._loading);
+    this._loading.start();
+
+    const lyrics = await get_lyrics(new_lyrics, {
+      signal: this.controller.signal,
+    })
+      .catch((err) => {
+        if (err.name !== "AbortError") throw err;
+        return null;
```
Comment (DaPigGuy) [inline src/components/player/now-playing/details/lyrics.ts:109]: inconsistent with `src/components/player/now-playing/details/related.ts`, the `return` is a bit redundant though, so perhaps just get rid of it here to match

File: src/components/player/now-playing/details/switcher.ts [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+    // remove all children
+    let child: Gtk.Widget | null;
+    while (child = this.get_first_child()) child.unparent();
+
+    // create already existing pages
+    list_model_to_array(this.pages).forEach((page) => {
+      const button = this.create_switcher_button(page);
+      this.append(button);
+      return button;
```
Comment (DaPigGuy) [inline src/components/player/now-playing/details/switcher.ts:134]:
```suggestion
```

File: src/components/player/now-playing/details/switcher.ts [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    removed: number,
+    added: number,
+  ) {
+    // remove buttons for removed pages
+    const removed_buttons = list_model_to_array(this.buttons).splice(
+      position,
+      position + removed,
+      ...(
+        new Array(added)
+          .map((index) => {
```
Comment (DaPigGuy) [inline src/components/player/now-playing/details/switcher.ts:177]:
```suggestion
        new Array(added)
          .fill()
          .map((_, index) => {
```

The array constructor initializes an array of empty values, which are ignored when trying to map

image

File: src/components/player/full.ts [hunk 5]
```diff
@@ -62,33 +88,44 @@ export class FullPlayerView extends Gtk.ActionBar {
...
-    const song = this.player.queue.current?.object;
+    this.bind_property(
+      "details-stack",
+      this._switcher,
+      "details-stack",
+      GObject.BindingFlags.BIDIRECTIONAL,
+    );
 
-    if (song) {
-      this.show_song(song!);
```
Comment (DaPigGuy) [inline src/components/player/full.ts:73]: `show_song` can be removed in this file

File: src/components/player/mini.ts [hunk 6]
```diff
@@ -35,29 +35,57 @@ export class MiniPlayerView extends Gtk.Overlay {
     super();
 
     this.player = get_player();
-  }
 
-  song_changed() {
-    const song = this.player.queue.current?.object;
-    if (song) {
-      this.show_song(song!);
```
Comment (DaPigGuy) [inline src/components/player/mini.ts:43]: `show_song` can be removed here also

File: data/ui/window.blp [hunk 7]
```diff
@@ -7,47 +7,24 @@ template $MuzikaWindow : Adw.ApplicationWindow {
...
-    reveal-bottom-bars: false;
-    bottom-bar-style: raised;
-    content: Adw.ToastOverlay toast_overlay {
-      Stack main_stack {
-        transition-type: slide_up_down;
-        
-        StackPage {
-          name: "main";
-          child: Adw.NavigationSplitView split_view {
+  // this should toast overlay should update it's margin-bottom
```
[the remark this answers is not in the document]
Comment (DaPigGuy) [reply data/ui/window.blp:18]: also seems like this wasn't implemented

**Review by DaPigGuy — COMMENTED** [review]
The bottom sheet looks awesome, well done!

My only concern is the view tab switcher functioning as a toggle to show the "player" view. Do other apps do this? Seems confusing to me.

File: data/ui/window.blp (hunk 7 above)
[the remark this answers is not in the document]
Comment (vixalien) [reply data/ui/window.blp]: It was just implemented!!!

File: src/components/player/now-playing/details/switcher.ts [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+    added: number,
+  ) {
+    // remove buttons for removed pages
+    const removed_buttons = list_model_to_array(this.buttons).splice(
+      position,
+      position + removed,
+      ...(
+        new Array(added)
+          .fill(null)
+          .map((index) => {
```
Comment (DaPigGuy) [inline src/components/player/now-playing/details/switcher.ts:177]:
```suggestion
          .map((_, index) => {
```
first arg is value, second arg is index

Comment (vixalien) [reply src/components/player/now-playing/details/switcher.ts:177]: nice catch! thanks.

**Comment by vixalien** [comment]
> My only concern is the view tab switcher functioning as a toggle to show the "player" view. Do other apps do this? Seems confusing to me.

Yep, Apple Music on mobile does this too

File: src/window.ts [hunk 9]
```diff
@@ -299,13 +275,41 @@ export class Window extends Adw.ApplicationWindow {
...
             action.set_state(GLib.Variant.new_string(string));
           }
         },
       },
+      {
+        name: "now-playing",
+        state: "false",
+        change_state: (action, param) => {
+          if (!param) return;
+          if (param?.get_boolean()) {
```
Comment (DaPigGuy) [inline src/window.ts:290]: no need for optional chaining here because of previous line

File: src/layout/shell.ts [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  vfunc_add_child(
+    builder: Gtk.Builder,
+    child: GObject.Object,
+    type?: string | null | undefined,
+  ): void {
+    // this._multi_layout_view is only set after initializing
+    if (this._multi_layout_view && child instanceof Gtk.Widget) {
+    return this._multi_layout_view.set_child("content", child);
```
Comment (DaPigGuy) [inline src/layout/shell.ts:111]: indentation

**Review by DaPigGuy — COMMENTED** [review]
~~Looks like de9768191e7e37e5905a1c9d948f903f7a3bb702 broke CI :thinking:~~

Edit: gnome-nightly seems to have caused the breakage, same issue on `main`

**Comment by vixalien** [comment]
> gnome-nightly seems to have caused the breakage, same issue on `main`

Yeah, the latest `node20` extension seems to be broken. You can temporarily fix this by downgrading the GNOME SDK to an earlier version (and refrain from updating)

```
flatpak update --commit=f74d8ee57f6c55458bbd9f38093753d7fe39505eedb532ff2d286bb456b9b861 org.gnome.Sdk//master
````

**Pull request merged**